RFC: Generic member access for dyn Error trait objects (#2895)
rust-lang/rfcs: text/0000-dyn-error-generic-member-access.md

- Feature Name: Add fns for generic member access to dyn Error and the Error trait
- Start Date: 2020-04-01
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/2895)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

> This RFC proposes a pair of additions to the `Error` trait to support accessing
> generic forms of context from `dyn Error` trait objects, one method on the
> `Error` trait itself for returning references to members based on a given type
> id, and another fn implemented for `dyn Error` that uses a generic return type
> to get the type id to pass into the trait object's fn. These functions will act
> as a generalized version of `backtrace` and `source`, and would primarily be
> used during error reporting when rendering a chain of opaque errors.

Reviewer: ```suggestion
This RFC proposes two additions to the `Error` trait to support accessing
generic forms of context from `dyn Error` trait objects, generalizing the
pattern used in `backtrace` and `source` and allowing ecosystem iteration
on error reporting infrastructure outside of the standard library. The two
proposed additions are a new trait method `Error::provide_context` which
offers `TypeId`-based member lookup and a new inherent fn
`<dyn Error>::context` which makes use of an implementor's `provide_context`
to return a typed reference directly.

These additions would primarily be useful in "error reporting" contexts where we
typically no longer have type information and may be composing errors from many
sources.
```

IMO the definitions themselves are very readable as an "intro" to what you're proposing if you put them together:

```rust
pub trait Error {
    /// Provide an untyped reference to a member whose type matches the provided `TypeId`.
    ///
    /// Returns `None` by default, implementors are encouraged to override.
    fn provide_context(&self, ty: TypeId) -> Option<&dyn Any> {
        None
    }
}

impl dyn Error {
    /// Retrieves a reference to `T`-typed context from the error if it is available.
    pub fn context<T: Any>(&self) -> Option<&T> {
        self.provide_context(TypeId::of::<T>())?.downcast_ref::<T>()
    }
}
```

    }
}
```

# Motivation
[motivation]: #motivation

Today, there are a number of forms of context that are traditionally gathered
when creating errors. These members are gathered so that a final error
reporting type or function can access them and render them independently of the
`Display` implementation for each specific error type. This allows for
consistently formatted and flexible error reports. Today, there are 2 such
forms of context that are traditionally gathered, `backtrace` and `source`.

However, the current approach of promoting each form of context to a fn on the
`Error` trait doesn't leave room for forms of context that are not commonly
used, or forms of context that are defined outside of the standard library.

## Example use cases this enables

* using `backtrace::Backtrace` instead of `std::backtrace::Backtrace`
* zig-like Error Return Traces by extracting `Location` types from errors
  gathered via `#[track_caller]` or some similar mechanism.
* error source trees instead of chains by accessing the source of an error as a
  slice of errors rather than as a single error, such as a set of errors caused
  when parsing a file

> * `SpanTrace` a backtrace like type from the `tracing-error` library

Author: I think that might be more detail than is necessary for the RFC, I've merged it with the "alternatives to std::backtrace" bullet and it has a docs.rs link

* Help text such as suggestions or warnings attached to an error report

By adding a generic form of these functions that works around the restriction
on generics in vtables we could support a greater diversity of error handling
needs and make room for experimentation with new forms of context in error
reports.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Error handling in rust consists mainly of two steps, creation/propogation and

> reporting. The `std::error::Error` trait exists to bridge the gap between these

Author: Not that I can think of, the only other purpose it has is downcasting, so if you wanted a type that could contain an open set of errors where the only thing you do with them is downcast them back to concrete types you'd just use `dyn Any`.

two steps. It does so by acting as a consistent interface that all error types
can implement to allow error reporting types to handle them in a consistent
manner when constructing reports for end users.

The error trait accomplishes this by providing a set of methods for accessing
members of `dyn Error` trait objects. The main member, the error message
itself, is handled by the `Display` trait which is a requirement for
implementing the Error trait. For accessing `dyn Error` members it provides the
`source` function, which conventionally represents the lower level error that
caused the current error. And for accessing a `Backtrace` of the state of the
stack when an error was created it provides the `backtrace` function. For all
other forms of context relevant to an Error Report the error trait provides the
`context`/`provide_context` functions.

As an example of how to use these types to construct an error report lets
explore how one could implement an error reporting type that retrieves the

> Location where each error in the chain was created, if it exists, and renders

Reviewer: nit: backticks for Location

> it as part of the chain of errors. Our end goal is to get an error report that

Reviewer: I'd start the walkthrough with the motivating output, much better headline

Author: Do you mean move this to the top of the guide level explanation? or to motivation / summary?

looks something like this:

```
Error:
    0: Failed to read instrs from ./path/to/instrs.json
        at instrs.rs:42
    1: No such file or directory (os error 2)
```

The first step is to define or use a Location type. In this example we will

> define our own but we could use also use `std::panic::Location` for example.

Reviewer: Ah, I might be the only person in the world who finds this really confusing but this really tripped me up (I've been staring at core::panic::Location for months)

Reviewer: Specifically what tripped me up was ending this sentence referencing std but then still defining our own. I'd pick one or the other and run with it.

        if id == TypeId::of::<Location>() {
            Some(&self.location)
        } else {
            None
        }
    }
}
```

And finally, we create an error reporter that prints the error and its source
recursively along with the location data if it was gathered.

```rust
struct ErrorReporter(Box<dyn Error + Send + Sync + 'static>);

impl fmt::Debug for ErrorReporter {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut current_error = Some(self.0.as_ref());
        let mut ind = 0;

        while let Some(error) = current_error {
            writeln!(fmt, "    {}: {}", ind, error)?;

            if let Some(location) = error.context::<Location>() {
                writeln!(fmt, "        at {}:{}", location.file, location.line)?;
            }

            ind += 1;
            current_error = error.source();
        }

        Ok(())
    }
}
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

There are two additions necessary to the standard library to implement this
proposal:

Add a function for dyn Error trait objects that will be used by error

> reporters to access members given a generic type. This function circumvents

Reviewer: s/generic//

restrictions on generics in trait functions by being implemented for trait
objects only, rather than as a member of the trait itself.

```rust
impl dyn Error {
    pub fn context<T: Any>(&self) -> Option<&T> {
        self.provide_context(TypeId::of::<T>())?.downcast_ref::<T>()
    }
}
```

With the expected usage:

```rust
// With explicit parameter passing
let spantrace = error.context::<SpanTrace>();

// With a type inference
fn get_spantrace(error: &(dyn Error + 'static)) -> Option<&SpanTrace> {
    error.context()
}
```

Add a member to the `Error` trait to provide the `&dyn Any` trait objects to
the `context` fn for each member based on the type_id.

```rust
trait Error {
    /// ...

>     fn provide_context(&self, id: TypeId) -> Option<&dyn Any> {

Reviewer: It occurs to me reading this that it seems a little more common to name this kind of thing `get_context` in Rust but IDK.

        None
    }
}
```

With the expected usage:

```rust
fn provide_context(&self, type_id: TypeId) -> Option<&dyn Any> {
    if id == TypeId::of::<Location>() {
        Some(&self.location)
    } else {
        None
    }
}
```

# Drawbacks
[drawbacks]: #drawbacks

* The API for defining how to return types is cumbersome and possibly not
  accessible for new rust users.
    * If the type is stored in an Option getting it converted to an `&Any` will
      probably challenge new devs, this can be made easier with documented
      examples covering common use cases and macros like `thiserror`.

> ```rust

Reviewer: Is this misplaced?

Author: no, but I wasn't really sure how to put a code snippet in a bullet list

} else if typeid == TypeId::of::<SpanTrace>() {
    self.span_trace.as_ref().map(|s| s as &dyn Any)
}
```

> * When you return the wrong type and the downcast fails you get `None` rather

Reviewer: this could be multiple sentences or have more punctuation or...

  than a compiler error guiding you to the right return type, which can make it
  challenging to debug mismatches between the type you return and the type you
  use to check against the type_id

>     * The downcast could be changed to panic when it fails

Reviewer: I'd call this an unknown rather than a drawback "how should the `context` fn handle failed downcasts?"

>     * There is an alternative implementation that mostly avoids this issue

Reviewer: ?

Author: Nika has a suggestion below, I really don't know which one is preferrable, they're both good, nika's has serious advantages, but it's also more complicated.

Author: Maybe I should just bite the bullet and make Nika's the primary suggestion and leave the `Any` version as the alternate, because its much shorter.

```rust
fn provide_context<'r, 'a>(&'a self, request: Request<'r, 'a>) -> ProvideResult<'r, 'a> {
    request
        .provide::<PathBuf>(&self.path)?
        .provide::<Path>(&self.path)?
        .provide::<dyn Debug>(&self.path)
}
```

The advantages of this design are that:

1. It supports accessing trait objects and slices
2. If the user specifies the type they are trying to pass in explicitly they
   will get compiler errors when the type doesn't match.
3. Takes advantage of deref sugar to help with conversions from wrapper types
   to inner types.
4. Less verbose implementation

The disadvatages are:

1. More verbose function signature, very lifetime heavy
2. The Request type uses unsafe code which needs to be verified
3. could encourage implementations where they pass the provider to
   `source.provide` first which would prevent the error reporter from knowing
   which error in the chain gathered each piece of context and might cause
   context to show up multiple times in a report.

# Prior art
[prior-art]: #prior-art

I do not know of any other languages whose error handling has similar
facilities for accessing members when reporting errors. For the most part prior

> art exists within rust itself in the form of previous additions to the `Error`

Reviewer: Java's Error casting can kinda do this, you can attempt to cast to the interface that has what you want and see what happens.

Author: I'll look into this, ty for the heads up

trait.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- What should the names of these functions be?
    - `context`/`context_ref`/`provide_context`
    - `member`/`member_ref`
    - `provide`/`request`
- Should we go with the implementation that uses `Any` or the one that supports

>   accessing dynamically sized types like traits and slices?

Reviewer: Another question I have: should a "probe a trait object by TypeId and/or Symbol" be a language feature in itself? i.e. you're proposing a constrained form of reflection (if you squint) -- what are the arguments against a more general form of it for any `dyn Trait`?

Author: :thinking:

- Should there be a by value version for accessing temporaries?
    - I bring this up specifically for the case where you want to use this
      function to get an `Option<&[&dyn Error]>` out of an error, in this case
      its unlikely that the error behind the trait object is actually storing
      the errors as `dyn Errors`, and theres no easy way to allocate storage to
      store the trait objects.

# Future possibilities
[future-possibilities]: #future-possibilities

> I'd love to see the various error creating libraries like `thiserror` adding

Reviewer: Prefer "the author" or "we" to "I" in something that'll become a normative text, IMO

support for making members exportable as context for reporters.

Also, I'm interested in adding support for `Error Return Traces`, similar to

> zigs, and I think that this accessor function might act as a critical piece of

Reviewer: It might be worth writing a blog post companion to this which sketches out the possibility?

Author: Good idea :D

- Feature Name: Add fns for generic member access to dyn Error and the Error trait
- Start Date: 2020-04-01
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/2895)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

This RFC proposes additions to the `Error` trait to support accessing generic
forms of context from `dyn Error` trait objects. This generalizes the pattern
used in `backtrace` and `source`. This proposal adds the method
`Error::get_context` to the error trait, which offers `TypeId`-based member
lookup, and a new inherent fn `<dyn Error>::context`, which makes use of an
implementor's `get_context` to return a typed reference directly. These
additions would primarily be useful in "error reporting" contexts, where we
typically no longer have type information and may be composing errors from many
sources.

> _note_: This RFC focuses on the more complicate of it's two proposed solutions

Reviewer 2: ```suggestion
_note_: This RFC focuses on the more complicated of it's two proposed solutions
```

- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/2895)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

This RFC proposes additions to the `Error` trait to support accessing generic
forms of context from `dyn Error` trait objects. This generalizes the pattern
used in `backtrace` and `source`. This proposal adds the method
`Error::get_context` to the `Error` trait, which offers `TypeId`-based member
lookup and a new inherent function `<dyn Error>::context` which makes use of an
implementor's `get_context` to return a typed reference directly. These
additions would primarily be useful for error reporting, where we typically no
longer have type information and may be composing errors from many sources.

_note_: This RFC focuses on the more complicated of it's two proposed
solutions. The proposed solution provides support accessing dynamically sized
types. The [alternative proposal] is easier to understand and may be more
palatable.

## TLDR

Add this method to the `Error` trait

```rust
pub trait Error {
    // ...

    /// Provides an object of type `T` in response to this request.

>     fn get_context<'r, 'a>(&'a self, request: Request<'r, 'a>) -> ProvideResult<'r, 'a> {

Reviewer 3: Ref https://github.com/rust-lang/rfcs/pull/2895#discussion_r403736887 @anp,

I find it strange that we have both `context()` function (a getter, sort of) and a `get_context()` function (sounds like a getter, but it's not). Keeping it `provide_context()` maybe better.

    /// `Some(&T)` if the request was fulfilled, and `None` otherwise.
    ///
    /// The `ObjectProviderExt` trait provides helper methods specifically for
    /// types implementing `ObjectProvider`.
    pub fn with<T: ?Sized + 'static, F>(f: F) -> Option<&'out T>
    where
        for<'a> F: FnOnce(Request<'a, 'out>) -> ProvideResult<'a, 'out>,
    {
        let mut buf = RequestBuf {
            type_id: TypeId::of::<T>(),
            value: None,
        };
        let _ = f(Request {
            buf: unsafe {
                NonNull::new_unchecked(&mut buf as *mut RequestBuf<'out, T> as *mut TypeId)
            },
            _marker: PhantomData,
        });
        buf.value
    }
}

// Needs to have a known layout so we can do unsafe pointer shenanigans.
#[repr(C)]
struct RequestBuf<'a, T: ?Sized> {
    type_id: TypeId,
    value: Option<&'a T>,
}

/// Marker type indicating a request has been fulfilled.
pub struct FulfilledRequest(PhantomData<&'static Cell<()>>);

/// Provider method return type.
///
/// Either `Ok(Request)` for an unfulfilled request, or `Err(FulfilledRequest)`
/// if the request was fulfilled.

> pub type ProvideResult<'r, 'a> = Result<Request<'r, 'a>, FulfilledRequest>;

Reviewer 3: This looks really odd that `Err` is the "happy case" especially as a public type. If the purpose is just to use `?` chaining it is probably better to just create an `impl Try` type like [LoopState](https://github.com/rust-lang/rust/blob/2454a68cfbb63aa7b8e09fe05114d5f98b2f9740/src/libcore/iter/mod.rs#L375).

```rust
enum Answer<T, S> {
    Found(T),
    Continue(S),
}
```

and, more recently, backtraces.

However, the current approach of promoting each form of context to a method on
the `Error` trait doesn't leave room for forms of context that are not commonly
used, or forms of context that are defined outside of the standard library.

## Extracting non-std types from `dyn Errors`

By adding a generic form of these member access functions we are no longer
restricted to types defined in the standard library. This opens the door to
many new forms of error reporting.

### Example use cases this enables

* using alternatives to `std::backtrace::Backtrace` such as
  `backtrace::Backtrace` or [`SpanTrace`]
* zig-like Error Return Traces by extracting `Location` types from errors
  gathered via `#[track_caller]` or similar.
* error source trees instead of chains by accessing the source of an error as a
  slice of errors rather than as a single error, such as a set of errors caused
  when parsing a file TODO reword
* Help text such as suggestions or warnings attached to an error report

## Moving `Error` into `libcore`

Adding a generic member access function to the `Error` trait and removing the

> `backtrace` function would make it possible to move the `Error` trait to libcore

Reviewer: I had forgotten that `Error::backtrace` was still unstable -- might be worth making that explicit.

without losing support for backtraces on std. The only difference being that
in places where you can currently write `error.backtrace()` on nightly you
would instead need to write `error.context::<Backtrace>()`.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Error handling in Rust consists of three steps: creation/propagation, handling,
and reporting. The `std::error::Error` trait exists to bridge the gap between

> creation and reporting. It does so by acting as a interface that all error

Reviewer: ```suggestion
creation and reporting. It does so by acting as an interface that all error
```

types can implement that defines how to access context intended for error
reports, such as the error message, source, or location it was created. This
allows error reporting types to handle errors in a consistent manner when
constructing reports for end users while still retaining control over the
format of the full report.

The `Error` trait accomplishes this by providing a set of methods for accessing
members of `dyn Error` trait objects. It requires that types implement the

> display trait, which acts as the interface to the main member, the error

Reviewer: ```suggestion
`Display` trait, which acts as the interface to the main member, the error
```

message itself.  It provides the `source` function for accessing `dyn Error`
members, which typically represent the current error's cause. It provides the

> `backtrace` function, for accessing a `Backtrace` of the state of the stack

Reviewer: Is it worth mentioning that backtrace is unstable?

Reviewer: FWIW realizing that this could subsume `Error::backtrace` is waaaay cool.

Author: it could also subsume `source` fwiw :D

Reviewer: True, but we can't *delete* `source`. I guess deprecation markers have been pretty successful in the ecosystem.

Author: yea, i don't really think its necessary to delete source, at a certain point it is okay to optimize your API for the happy path, and sources are by far the most common kind of context, and its only adding a dependency on its own definition, so its presence not an issue as far as moving the Error trait around is concerned afaik.

            },
        }
    })
}
```

Then, we need to implement the `Error` trait to expose these members to the error reporter.

```rust
impl std::error::Error for ExampleError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        Some(&self.source)
    }

    fn provide_context<'a>(&'a self, mut request: Pin<&mut Request<'a>>) {
        request.provide::<Location>(&self.location);
    }
}
```

And, finally, we create an error reporter that prints the error and its source
recursively, along with any location data that was gathered

```rust
struct ErrorReporter(Box<dyn Error + Send + Sync + 'static>);

impl fmt::Debug for ErrorReporter {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut current_error = Some(self.0.as_ref());
        let mut ind = 0;

        while let Some(error) = current_error {
            writeln!(fmt, "    {}: {}", ind, error)?;

            if let Some(location) = error.context::<Location>() {
                writeln!(fmt, "        at {}:{}", location.file, location.line)?;
            }

            ind += 1;
            current_error = error.source();
        }

        Ok(())
    }
}
```

> As you can see the `Error` trait provides the facilities needed to create error

Reviewer: I think you could delete this sentence without losing your point, or maybe make it more specific, referencing the specific portions of the example that highlight this?

Author: I just feel like I need some sort of conclusion, otherwise the guide ends super suddenly and it feels bad. Coming up with a good conclusion is proving difficult :(

Reviewer: What if you move the example output from above down here and make the "intended output" above a bit more generic?

```
Error:
    0: ERROR MESSAGE
        at LOCATION
    1: PRIOR MESSAGE
    ...
```

reports enriched by information that may be present in source errors.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

The following changes need to be made to implement this proposal:

> ### Add a type like [`Request`] to core

Reviewer: Is it possible to describe the type in terms of its role, rather than in terms of another type?

"Add a `Request` type to core for providing type-indexed context"? IDK

> This type fills the same role as `&dyn Any` except that it supports other trait

Reviewer: This seems like an addendum to a more "central" description of `Request`, which I imagine could take the form of Any's docs: "A trait to emulate dynamic typing."

What about "`Request` is a type to emulate nested dynamic typing." ?

objects as the requested type.

Here is the implementation for the proof of concept, based on Nika Layzell's
[object-provider crate]:

> A usable version of this is available in the [proof of concept] repo under

Reviewer: Would a WIP PR with a couple example tests be too much? Seems like the next step up from having made a functioning proof-of-concept. IIRC there are a few RFCs where a "companion" PR was uploaded to illustrate a PoC of an approach.

Author: oh thats a good idea. It wouldn't be able to show the "move to core" bits for the error trait because theres the minor issue of `downcast` having `Box` in its API and needing to be in the same crate that defines `Error`, but the rest of it could be done.

One downside to that though is its a little harder to pull and use, because presumably that would be a branch of `rust-lang/rust`, so the compile times would be a lot worse and you'd have to know how to use `x.py`, which isn't hard but its definitely _harder_.

I'll probably do it though, but still treat the proof-of-concept repo as the source of truth where changes are initially made.

    ///
    /// This method can be chained within `provide` implementations to concisely
    /// provide multiple objects.
    pub fn provide<T: ?Sized + 'static>(self: Pin<&mut Self>, value: &'a T) -> Pin<&mut Self> {
        self.provide_with(|| value)
    }

    /// Lazily provides an object of type `T` in response to this request.
    ///
    /// If an object of type `T` has already been provided for this request, the
    /// existing value will be replaced by the newly provided value.
    ///
    /// The passed closure is only called if the value will be successfully
    /// provided.
    ///
    /// This method can be chained within `provide` implementations to concisely
    /// provide multiple objects.
    pub fn provide_with<T: ?Sized + 'static, F>(mut self: Pin<&mut Self>, cb: F) -> Pin<&mut Self>
    where
        F: FnOnce() -> &'a T,
    {
        if let Some(buf) = self.as_mut().downcast_buf::<T>() {
            // NOTE: We could've already provided a value here of type `T`,
            // which will be clobbered in this case.
            *buf = Some(cb());
        }
        self
    }

    /// Get the `TypeId` of the requested type.

>     pub fn type_id(&self) -> TypeId {

Reviewer: If your goal is to use explicit type parameters as the public API to avoid the hazards of manually-compared TypeIds, should this still be public?

Author: This is an open question but I lean towards no, I'll add it to the unresolved questions section.

Reviewer 4: I included `type_id` and `is` in the API largely to improve the flexibility of the API for use-cases other than the `Error` trait.

A use-case which would be difficult without the `type_id` and `is` methods would be one where an interface is optionally provided, but the decision of whether or not to provide it is expensive. The `provide_with` method is required to return `&T`. With `is`, it would be possible to write code like:

```rust
if req.is::<T>() && expensive_available_check() {
    req.provide::<T>(&whatever);
}
```

Theoretically we could provide something like `try_provide_with` which takes `FnOnce() -> Option<&T>`, but generally control-flow within closures is much more limited, so providing options seems good.

If we really don't want to allow `is` and `type_id`, we could potentially expose something similar to the `downcast_buf` method (probably with a better name/API) which would ensure the types match, but allow non-closure-based control flow. Perhaps something like this:

```rust
impl Request<'a> {
    fn downcast<T: ?Sized + 'static>(&mut self) -> Option<&mut Option<&'a T>>;
}
```

It might be a good idea to not stabilize any low level APIs in the first pass, and revisit these later as needs arise.

Reviewer: This makes sense to me - I was actually wondering about being able to hoist the branch into the implementor code just like your example. Agreed it seems less important for the first pass.

Does making `is<T>()` public require doing the same for `type_id()`? The former seems obviously useful and hard to mess up, the latter seems like the actual source of footguns.

Reviewer 4: It would be totally possible/reasonable to only expose `is<T>()`, and not expose `type_id()` if we're worried that it won't be used correctly

            .finish()
    }
}

// Needs to have a known layout so we can do unsafe pointer shenanigans.
#[repr(C)]
struct RequestBuf<'a, T: ?Sized + 'static> {
    request: Request<'a>,
    value: Option<&'a T>,
}

impl<'a, T: ?Sized + 'static> RequestBuf<'a, T> {
    fn new() -> Self {
        RequestBuf {
            request: Request {
                type_id: TypeId::of::<T>(),
                _pinned: PhantomPinned,
                _marker: PhantomData,
            },
            value: None,
        }
    }

    fn request(self: Pin<&mut Self>) -> Pin<&mut Request<'a>> {
        // safety: projecting Pin onto our `request` field.
        unsafe { self.map_unchecked_mut(|this| &mut this.request) }
    }

    fn take(self: Pin<&mut Self>) -> Option<&'a T> {
        // safety: `Option<&'a T>` is `Unpin`
        unsafe { self.get_unchecked_mut().value.take() }
    }
}
```

### Define a generic accessor on the `Error` trait

```rust
pub trait Error {
    // ...

    fn provide_context<'a>(&'a self, _request: Pin<&mut Request<'a>>) {}
}
```

### Use this `Request` type to handle passing generic types out of the trait object

```rust
impl dyn Error {

>     pub fn context<T: ?Sized + 'static>(&self) -> Option<&T> {

Reviewer: I think it would be really useful to have this in the TLDR

trait whenever a new type reaches critical levels of popularity in error
reporting.

If we choose to do nothing we will continue to see hacks around the current
limitations on the `Error` trait such as the `Fail` trait, which added the
missing function access methods that didn't previously exist on the `Error`
trait and type erasure / unnecessary boxing of errors to enable downcasting to
extract members.
[[1]](https://docs.rs/tracing-error/0.1.2/src/tracing_error/error.rs.html#269-274).

## Use an alternative proposal that relies on the `Any` trait for downcasting

```rust
pub trait Error {
    /// Provide an untyped reference to a member whose type matches the provided `TypeId`.
    ///
    /// Returns `None` by default, implementors are encouraged to override.
    fn provide_context(&self, ty: TypeId) -> Option<&dyn Any> {
        None
    }
}

impl dyn Error {
    /// Retrieve a reference to `T`-typed context from the error if it is available.
    pub fn context<T: Any>(&self) -> Option<&T> {
        self.provide_context(TypeId::of::<T>())?.downcast_ref::<T>()
    }
}
```

### Why isn't this the primary proposal?

> There are two big issues with using the `Any` trait that I believe justify the

Reviewer: ```suggestion
There are two significant issues with using the `Any` trait that motivate the
```

more complicated solution.

- You cannot return dynamically sized types as `&dyn Any`
- It's easy to introduce runtime errors with `&dyn Any` by either comparing to
  or returning the wrong type

> By making all the type id comparison internal to the `Request` type it is

Reviewer: ```suggestion
By making all the `TypeId` comparison internal to the `Request` type it is
```

> impossible to compare the wrong type ids. By encouraging explicit type

Reviewer: ```suggestion
impossible to compare the wrong `TypeId`s. By encouraging explicit type
```

parameters when calling `provide` the compiler is able to catch errors where
the type passed in doesn't match the type that was expected. So while the API
for the main proposal is more complicated it should be less error prone.

# Prior art
[prior-art]: #prior-art

I do not know of any other languages whose error handling has similar

> facilities for accessing members when reporting errors. For the most part,

Reviewer: AIUI the analogous operation in a class-inheritance language like Java would be to attempt a cast to a subclass, or am I missing something? This seems like the mix-in version of that if I squint.

Author: yea. This could probably be used for upcasting too, you just have to return `&self`, I'm not really sure if thats prior art but ill see about mentioning similarities.

Reviewer: Yeah it’s more “prior niche fulfillment” than “prior art”.

prior art for this proposal comes from within rust itself in the form of
previous additions to the `Error` trait.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> * What should the names of these functions be?

Reviewer: I'm most interested in the question raised above about `std::Error::context` conflicting with `failure::Fail::context` and `anyhow::Error::context`. Have you considered a different name for the "reporter-facing" API like `Error::from_context` or `Error::request_context`?

Author: I'm not sure I like `from_context`, but `request_context` is a good option. Though I think I'd rather conflict honestly, to encourage downstream error libraries to change their API to match the mental model for the error trait that is proposed here. In those libraries they view the entire chain of errors as one error, where each time you wrap it you're adding context. This isn't consistent with the error trait though, with the API provided here each error is distinct and the saved sources are context. So while its definitely true that errors in a chain have a context-like relationship a bunch of the error handling libraries out there model it backwards and I think that's a source of confusion in the ecosystem and part of why people complain about error handling in rust so much.

    * `context`/`context_ref`/`provide_context`/`provide_context`
    * `member`/`member_ref`
    * `provide`/`request`
* Should there be a by-value version for accessing temporaries?
    * We bring this up specifically for the case where you want to use this
      function to get an `Option<&[&dyn Error]>` out of an error, in this case,
      it is unlikely that the error behind the trait object is actually storing
      the errors as `dyn Error`s, and theres no easy way to allocate storage to
      store the trait objects.
* How should context handle failed downcasts?

>     * suggestion: panic, as providing a type that doesn't match the typeid

Reviewer: What's the alternative? Map it to `None` as if no context was provided?

Author: The problem isn't when they return None, its when they return the wrong type

```rust
if TypeId::of::<SpanTrace>() == id {
   Some(&self.backtrace as &dyn Any)
}
```

This is pretty clearly a program error imo, this is only a problem with the `Any` based approach though, so I should probably remove this one. With `Request` your only options are to either do `request.provide::<SpanTrace>(&self.backtrace)` which is caught at compile time or do `request.provide(&self.backtrace)` which will just correctly compare the typeID to a `Backtrace`, which might still be wrong if you were actually trying to return a SpanTrace, like if you do `&self.backtrace` twice and meant to do `&self.span_trace` the second time. It should probably panic if it is provided the same type twice.

        if self.is::<T>() {
            // Safety: `self` is pinned, meaning it exists as the first
            // field within our `RequestBuf`. As the type matches, and
            // `RequestBuf` has a known in-memory layout, this downcast is
            // sound.
            unsafe {
                let ptr = self.get_unchecked_mut() as *mut Self as *mut RequestBuf<'a, T>;
                Some(&mut (*ptr).value)
            }
        } else {
            None
        }
    }

    /// Calls the provided closure with a request for the the type `T`, returning
    /// `Some(&T)` if the request was fulfilled, and `None` otherwise.
    pub fn with<T: ?Sized + 'static, F>(f: F) -> Option<&'a T>
    where
        F: FnOnce(Pin<&mut Self>),
    {
        let mut buf = RequestBuf::new();
        // safety: We never move `buf` after creating `pinned`.
        let mut pinned = unsafe { Pin::new_unchecked(&mut buf) };
        f(pinned.as_mut().request());
        pinned.take()
    }
}

impl<'a> fmt::Debug for Request<'a> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.debug_struct("Request")
            .field("type_id", &self.type_id)
            .finish()
    }
}

> // Needs to have a known layout so we can do unsafe pointer shenanigans.

Reviewer 5: If rust had an offset_of operator or macro, I don't think `#[repr(C)]` would be necessary here.

Reviewer 6: There is [`memoffset`](https://lib.rs/crates/memoffset) crate. AFAIU it will work on stable in sound way pretty soon, when `std::ptr::raw_ref` is stable.

                line: line!(),
            },
        }
    })
}
```

Then, we need to implement the `Error` trait to expose these members to the error reporter.

```rust
impl std::error::Error for ExampleError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        Some(&self.source)
    }

    fn provide_context<'a>(&'a self, mut request: Pin<&mut Request<'a>>) {
        request.provide::<Location>(&self.location);
    }
}
```

And, finally, we create an error reporter that prints the error and its source
recursively, along with any location data that was gathered

```rust
struct ErrorReporter(Box<dyn Error + Send + Sync + 'static>);

impl fmt::Debug for ErrorReporter {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut current_error = Some(self.0.as_ref());
        let mut ind = 0;

        while let Some(error) = current_error {
            writeln!(fmt, "    {}: {}", ind, error)?;

            if let Some(location) = error.context::<Location>() {
                writeln!(fmt, "        at {}:{}", location.file, location.line)?;
            }

            ind += 1;
            current_error = error.source();

>         }

Reviewer 7: That loop is a great use case for using `self.0.chain().enumerate()` (still unstable, cc rust-lang/rust#58520).

Author: This can also be done pretty well with `iter::successors(some(self.0.as_ref()), |s| s.source()).enumerate();`

  when parsing a file TODO reword
* Help text such as suggestions or warnings attached to an error report

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Error handling in Rust consists of three main steps: creation/propagation,
handling, and reporting. The `std::error::Error` trait exists to bridge the
gap between creation and reporting. It does so by acting as an interface that
all error types can implement that defines how to access context intended for
error reports, such as the error message, source, or location it was created.
This allows error reporting types to handle errors in a consistent manner
when constructing reports for end users while still retaining control over
the format of the full report.

The `Error` trait accomplishes this by providing a set of methods for accessing
members of `dyn Error` trait objects. It requires that types implement the
`Display` trait, which acts as the interface to the main member, the error
message itself.  It provides the `source` function for accessing `dyn Error`
members, which typically represent the current error's cause. Via

> `#![feature(backtrace)]` it provides the `backtrace` function, for accessing a

Reviewer 8: This seems outdated. In master branch I don't see any `backtrace` function on the `Error` trait with or without that feature. Maybe the `backtrace` feature was already merged but left out the `backtrace` method on `Error`. This RFC makes it unnecessary anyway so I'll look into updating this section here.

Reviewer 9: There used to be an unstable `backtrace` function on `Error`, but it was removed in favor for the generic member access proposed in this RFC.

        } else {
            None
        }
    }

    #[inline]
    #[cfg(feature = "alloc")]
    pub fn downcast_box<I>(self: Box<Self>) -> Result<Box<I::Type>, Box<Self>>
    where
        I: Tag<'a>,
    {
        if self.is::<I>() {
            unsafe {
                // SAFETY: Just checked whether we're pointing to a
                // `TaggedImpl<'a, I>`, which was cast to from an `I::Type`.
                let raw: *mut dyn Tagged<'a> = Box::into_raw(self);
                Ok(Box::from_raw(raw as *mut I::Type))
            }
        } else {
            Err(self)
        }
    }
}
```

### Define a generic accessor on the `Error` trait

```rust
pub trait Error {
    // ...

    fn provide_context<'a>(&'a self, _request: &mut Request<'a>) {}
}
```

### Use this `Request` type to handle passing generic types out of the trait object

```rust
impl dyn Error {
    pub fn context_ref<T: ?Sized + 'static>(&self) -> Option<&T> {
        Request::request_ref(|req| self.provide_context(req))
    }

    pub fn context<T: 'static>(&self) -> Option<T> {
        Request::request_value(|req| self.provide_context(req))
    }
}
```

> # Drawbacks

Reviewer 10: A potential additional drawback is that requesting types dynamically like this can lead to surprising changes in runtime behavior when changing versions of (transitive) dependencies.

Given:

* A crate `foo` with versions 1.0.0 and 2.0.0, each of which provides a type `Foo`.
* A crate `bar` which depends on version 1.0.0 of `foo` and whose errors provide `Foo`.
* A crate `baz` which depends `bar` and on version 2.0.0 of `foo` and tries to request `Foo` from `bar`'s errors.

Then a situation is created in which `baz` will fail to request `Foo` from `bar`'s error because the requested `Foo` and provided `Foo` are actually distinct types because the transitive dependency `foo`'s versions don't match up.

This could even happen in a semver-compatible release of `bar` and `baz` because technically the API is not broken: the type can still be requested, but it will simply begin returning `None` instead of `Some(_)`. This would be even worse if `foo`, `bar,` and `baz` are all library crates depended on by an application crate and the semver-incompatible versions of `foo` are introduced by a semver-compatible change in `bar` or `baz` which would get pulled in by running `cargo update` on the application crate. The application crate has no recourse in this situation aside from waiting for `bar` and `baz` to agree on a semver-compatible version of `foo` again.

(I've been playing with an alternate design based on an associated type rather than generic member access which causes this situation to be a compile time error instead of a runtime behavior change, but that design requires traits with lifetime GATs to be dyn safe to be actually good, and associated type defaults for it to *maybe* be possible to retrofit that design onto the standard library's existing error trait.)

Author: I agree this is a footgun I'd prefer to avoid but as far as I can tell this is not a problem being uniquely introduced by generic member access. This is a problem with type erasure in general and the same situation could be encountered by having foo return `Foo` as a `Box<dyn Any>` in any of it's public API's. If `baz` tries to downcast using it's own path to `foo::Foo` it will run into the same type ID mismatch.

I avoid this in my crates currently by re-exporting types from crates that are part of my public API (https://docs.rs/color-eyre/latest/color_eyre/#reexports). Applying this to your example you could avoid this footgun if you downcast/request `Foo` via `bar::foo::Foo` which would correctly resolve to whatever version it is using. This of course gets a little silly when you want to be able to handle all the versions of `Foo`, because then presumably you have to try to request `Foo` once for each path through which you could possibly reach it, which is definitely a subpar user experience, so not necessarily a great solution.

That said, I was never able to figure out how to implement generic member access w/o type erasure and runtime checks. I'm curious to understand more about your potential alternative design, because I would definitely love to see a better solution but I don't see how it is possible given the constraints of the error trait and the goals of generic member access.

Reviewer 10: > I agree this is a footgun I'd prefer to avoid but as far as I can tell this is not a problem being uniquely introduced by generic member access.

True.

> This is a problem with type erasure in general and the same situation could be encountered by having foo return `Foo` as a `Box<dyn Any>` in any of it's public API's.

However, it seems to me that this kind of thing is relatively rare currently, both in the standard library and ecosystem crates. The only standard library exception I can think of is `std::io::Error`, for which I found (part of?) the rationale for in https://github.com/rust-lang/rust/pull/24793#issuecomment-97153138, and I haven't personally encountered any ecosystem crates that make use of this and tell users in documentation that they can/should downcast the inner error to some concrete type.

So, I think making type erasure a central part of the error handling API of the standard library would probably be the main place this kind of problem would come up. I'm not sure the existing occasional cases of downcasting in public APIs can be used to justify a standard library API built entirely on downcasting.

> Applying this to your example you could avoid this footgun if you downcast/request `Foo` via `bar::foo::Foo` which would correctly resolve to whatever version it is using.

While true, I imagine another place this might become a problem is in a library whose purpose is to use generic member access to generate pretty error reports. This hypothetical library might hypothetically pull in the `backtrace` crate for handling backtraces. If any other crate in the dependency tree has errors that provide a backtrace from a different version of the `backtrace` crate, the hypothetical library will not be able to extract them. I don't think this can be solved with re-exports, because that would force the library to have a re-export of every crate that has errors that provide a `backtrace` crate backtrace.

Your next point about requesting the type through each path (including crate version) does provide a solution there, but also as you point out, would not be the most fun to maintain.

> I'm curious to understand more about your potential alternative design, because I would definitely love to see a better solution but I don't see how it is possible given the constraints of the error trait and the goals of generic member access.

I can't promise that this particular idea is any better, but for what it's worth, it goes like this:

```rust
trait Error {
    /// Type that contains details about this error.
    ///
    /// For example, this could be a struct containing a String with help text and a Backtrace.
    type Details<'a> = ()
        where Self: 'a;

    /// Get the details about this error.
    fn details(&self) -> Self::Details<'_>;
}
```

As written, this requires some language features that aren't implemented yet:

1. Associated type defaults. Necessary to make it possible to backport this onto `core::error::Error`, but luckily not required for defining a new error trait in a new crate.
2. `dyn`-safety for traits with lifetime GATs. In my current implementation, I've simply dropped the lifetime and am just returning an owned value, which has some overhead but at least allows experimenting with the rest of the design.

Another part of this design is a `map`-like function to change the type of `Details` when composing errors that have different concrete types.

To approximate generic member access to make it possible to implement a hypothetical library for pretty-printing errors like described above, I imagine a crate could define some traits like `trait GetBacktrace { fn get_backtrace(&self) -> Option<&Backtrace>; }` (or one trait with many methods, I suppose) and then adding these traits as bounds on `Error::Details` on the pretty-printing functions. The author of a binary crate could then implement these traits for their `Error::Details` type and use the library's pretty-printing code.

Reviewer 5: Wouldn't that still have a similar problem? To adapt your example from before:

- A crate foo with versions 1.0.0 and 2.0.0, each of which provides a trait Foo.
- A crate bar which depends on version 1.0.0 of foo and whose errors have a `Details` type that implements Foo 
- A crate baz which depends bar and on version 2.0.0 of foo and has uses a bound that requires `Error::Details` to implement Foo (from version 2.0.0 of foo)

baz still can't get the the `Foo` details from errors from bar.

Reviewer 10: That's correct, but the difference is *how* it can't get the details. With the generic member access design, this would fail at run time: the generic member access functions would return `None`. With the associated type design, this would fail at compile time: a compilation error would be emitted about the trait bound being unsatisfied.

Reviewer 5: You can currently have a `Box<dyn Error + 'static>` (or reference, or some other lifetime). Associated types and dyn-safety for traits with lifetime GATs isn't enough. If you had a `Details` type that isn't equal to the default, your error won't be compatible with anything that types a dynamic error, without some other kind of mechanism. And there may be some benefit to a compile time check. But that also means you lose use cases where you might want to extract data dynamically without knowing if it exists at runtime.

Reviewer 10: > You can currently have a `Box<dyn Error + 'static>` (or reference, or some other lifetime). Associated types and dyn-safety for traits with lifetime GATs isn't enough. If you had a `Details` type that isn't equal to the default, your error won't be compatible with anything that types a dynamic error, without some other kind of mechanism.

I guess a really important question is how many notable occurrences of `Box<dyn Error>` (and similar) are there in public APIs? The only one I can think of off the top of my head is `std::io::Error::{other, into_inner}`. In such a case, I believe the `map` function I alluded to before should be sufficient. To extract the original `Details`, you'd have to downcast to the concrete type first, which I think is not really a regression because that's how `std::io::Error` is currently intended to work. For internal-only APIs, the type could simply be changed to specify the desired `Details`.

> But that also means you lose use cases where you might want to extract data dynamically without knowing if it exists at runtime.

I suspect these cases could be covered for `Error`-processing functions by adding bounds on `Details` of traits that contain accessor functions that return `Option<T>` so that `Error` implementors have the, uh, option of returning `None`. Are there any situations where that would be insufficient?

Reviewer 5: > I guess a really important question is how many notable occurrences of Box<dyn Error> (and similar) are there in public APIs?

Well, I think returning a `Result<(), Box<dyn Error + 'static>` from main is a somewhat common pattern. And while you could maybe replace that with `Result<(), Box<dyn Error<Details=SomeType> + 'static>` , you would still need the same Details type for all errors, which gets messy if your errors come from different sources. 

And you have libraries like [anyhow](https://docs.rs/anyhow/latest/anyhow/) that make heavy use of `dyn core::error:Error`, such as in `root_cause()` and `chain()` and `from_boxed()` in anyhow. 

> To extract the original Details, you'd have to downcast to the concrete type first, which I think is not really a regression because that's how std::io::Error is currently intended to work

But that means you have to know what the concrete type is. And `Details` would need to implement a trait that has support for downcasting. And how do you handle combining multiple detail types into the a single error. Do you try casting to a struct that has the exact combination of details you expect? Do you use a collection of Details objects, and then try downcasting each one of those to what you expect?

Reviewer 10: > Well, I think returning a Result<(), Box<dyn Error + 'static> from main is a somewhat common pattern.

The return type of `main` is not a public API, and the requirements of the `Termination` trait allow you to choose other types for the `Err` variant, so this would be easy to change to specify a details type.

> And you have libraries like [anyhow](https://docs.rs/anyhow/latest/anyhow/) that make heavy use of `dyn core::error:Error`, such as in `root_cause()` and `chain()` and `from_boxed()` in anyhow.

I don't think this really counts towards what I was asking either. For `anyhow` (and similar crates) to be relevant here we'd need to know how often other library crates have `anyhow::Error` in their public interface.

> But that means you have to know what the concrete type is.

Worth noting that this is not a regression from the status quo.

> And how do you handle combining multiple detail types into the a single error. Do you try casting to a struct that has the exact combination of details you expect? Do you use a collection of Details objects, and then try downcasting each one of those to what you expect?

You would use the `map` function to compose detail types. You could implement your details type to do any of those things if you wanted, but I think a simpler approach would be to `impl From<OtherDetails> for MyDetails` or so.

the chain was created (if it captured a location) and render it as part of the
chain of errors. Our end goal is to get an error report that looks something
like this:

```
Error:
    0: ERROR MESSAGE
        at LOCATION
    1: SOURCE'S ERROR MESSAGE
        at SOURCE'S LOCATION
    2: SOURCE'S SOURCE'S ERROR MESSAGE
    ...
```

The first step is to define or use a type to represent a source location. In
this example, we will define our own:

```rust
struct Location {
    file: &'static str,
    line: usize,
}
```

Next, we need to gather the location when creating our error types.

```rust
struct ExampleError {
    source: std::io::Error,
    location: Location,
    path: PathBuf,
}

impl fmt::Display for ExampleError {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(fmt, "Failed to read instrs from {}", path.display())
    }
}

fn read_instrs(path: &Path) -> Result<String, ExampleError> {
    std::fs::read_to_string(path).map_err(|source| {
        ExampleError {
            source,
            path: path.to_owned(),
            location: Location {
                file: file!(),

>                 line: line!(),

Author: correct. It is already possible to do this with `track_caller`, so we could update this to include `track_caller`. I don't recall if there was a specific reason I avoided using track_caller here but I don't mind updating this example.

struct RefResponse<'a, T: ?Sized + 'static>(Option<&'a T>);
impl<'a, T: ?Sized + 'static> private::Response<'a> for RefResponse<'a, T> {}

/// A response to a value request.
struct ValueResponse<T: 'static>(Option<T>);
impl<'a, T: 'static> private::Response<'a> for ValueResponse<T> {}

/// A dynamic request for an object based on its type.
pub struct Request<'a, R = dyn private::Response<'a>>
where
    R: ?Sized + private::Response<'a>,
{
    marker: PhantomData<&'a ()>,

    /// A `TypeId` marker for the type stored in `R`.
    ///
    /// Will be the TypeId of either `RefResponse<'static, T>` or
    /// `ValueResponse<T>`.
    type_id: TypeId,

    /// A type erased `RefResponse` or `ValueResponse` containing the response
    /// value.
    response: R,
}

impl<'a> Request<'a> {
    /// Perform a checked downcast of `response` to `Option<&'a T>`
    fn downcast_ref_response<'b, T: ?Sized + 'static>(
        &'b mut self,
    ) -> Option<&'b mut RefResponse<'a, T>> {
        if self.is_ref::<T>() {
            // safety: If `self.is_ref::<T>()` returns true, `response` must be
            // of the correct type. This is enforced by the private `type_id`
            // field.

>             Some(unsafe { &mut *(&mut self.response as *mut _ as *mut RefResponse<'a, T>) })

Reviewer 11: Maybe my coercion-fu is lacking, but what is the `as *mut _` part in the middle achieving?

Reviewer 12: Mutable references (in this case `&mut R`) can only be cast to pointers to the same type. Here, that pointer is then cast from `*mut R` to `*mut RefResponse<'a, T>`.

Author: All of this is out of date anyways, I need to update it to use https://docs.rs/dyno/ as the example

    /// ignored.
    ///
    /// The passed closure is only called if the value will be successfully
    /// provided.
    ///
    /// This method can be chained within `provide` implementations to concisely
    /// provide multiple objects.
    ///
    /// # Example
    ///
    /// ```
    /// # use object_provider::{ObjectProvider, Request};
    /// # fn expensive_condition() -> bool { true }
    /// struct MyProvider {
    ///     a: String,
    ///     b: String,
    /// }
    ///
    /// impl ObjectProvider for MyProvider {
    ///     fn provide<'a>(&'a self, request: &mut Request<'a>) {
    ///         request.provide_ref_with::<String, _>(|| {
    ///             if expensive_condition() {
    ///                 &self.a
    ///             } else {
    ///                 &self.b
    ///             }
    ///         });
    ///     }
    /// }
    /// ```
    pub fn provide_ref_with<T: ?Sized + 'static, F>(&mut self, cb: F) -> &mut Self
    where
        F: FnOnce() -> &'a T,
    {
        if let Some(RefResponse(response @ None)) = self.downcast_ref_response::<T>() {
            *response = Some(cb());

>         }

Reviewer 11: Surely there's a typo here somewhere? `response @ None` can't be doing anything interesting, and it doesn't mean anything to try to deref it in the next line. Maybe the second line is meant to be `*self.response =`, but then I have no clue what the purpose of the `@` pattern is.

Reviewer 12: Does this help explain it?

```rust
struct Data(Option<()>);

let mut data = Data(None);
if let Some(Data(empty @ None)) = Some(&mut data) {
    *empty = Some(());
}
assert_eq!(data.0, Some(()));
```

Reviewer 11: Oof, the default binding modes threw me for a loop here. I thought they only worked at the top level of each pattern, it didn't occur to me that `response @ None` could be desugaring to `response @ &mut None`.

# Prior art
[prior-art]: #prior-art

I do not know of any other languages whose error handling has similar
facilities for accessing members when reporting errors. For the most part,
prior art for this proposal comes from within rust itself in the form of
previous additions to the `Error` trait.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

* What should the names of these functions be?
    * `context`/`context_ref`/`provide_context`/`provide_context`/`request_context`
    * `member`/`member_ref`
    * `provide`/`request`
* ~~Should there be a by-value version for accessing temporaries?~~ **Update**:
  The object provider API in this RFC has been updated to include a by-value
  variant for passing out owned data.
    * ~~We bring this up specifically for the case where you want to use this
      function to get an `Option<&[&dyn Error]>` out of an error, in this case,
      it is unlikely that the error behind the trait object is actually storing
      the errors as `dyn Error`s, and theres no easy way to allocate storage to
      store the trait objects.~~

# Future possibilities
[future-possibilities]: #future-possibilities

> This opens the door to supporting `Error Return Traces`, similar to zigs, where

Author: ```suggestion
This opens the door to supporting [`Error Return Traces`](https://ziglang.org/documentation/master/#toc-Error-Return-Traces), similar to zigs, where
```

[object-provider crate]:

```rust
use core::any::TypeId;
use core::cell::Cell;
use core::fmt;
use core::marker::PhantomData;
use core::pin::Pin;

/// A dynamic request for an object based on its type.
///
/// `'out` is the lifetime of the requested reference.
#[repr(transparent)]

> pub struct Request<'out>(RequestBuf<PhantomData<&'out Cell<()>>>);

Reviewer 3: need to include a `PhantomPinned` somewhere (maybe inside `RequestBuf` for simplicity), to avoid this type becoming `Unpin`.

// FIXME: The argument of the RequestBuf should be a thin unsized type,
// but `extern type` is impossible to use correctly right now
// (it cannot be placed at offset > 0, and it cannot be placed inside a union).
// Since miri doesn't complain we'll let it slide.

impl<'out> Request<'out> {
    /// Provides an object of type `T` in response to this request.
    ///
    /// Returns `Err(FulfilledRequest)` if the value was successfully provided,
    /// and `Ok(self)` if `T` was not the type being requested.
    ///
    /// This method can be chained within `provide` implementations using the
    /// `?` operator to concisely provide multiple objects.

>     pub fn provide<T: ?Sized + 'static>(&mut self, value: &'out T) -> &mut Self {

Reviewer 3: ```suggestion
    pub fn provide<T: ?Sized + 'static>(self: Pin<&mut Self>, value: &'out T) -> Pin<&mut Self> {
```

    ///
    /// This method can be chained within `provide` implementations using the
    /// `?` operator to concisely provide multiple objects.

>     pub fn provide_with<T: ?Sized + 'static, F>(&mut self, cb: F) -> &mut Self

Reviewer 3: ```suggestion
    pub fn provide_with<T: ?Sized + 'static, F>(self: Pin<&mut Self>, cb: F) -> Pin<&mut Self>
```

Author: this line no longer compiles when you apply this change.

```rust
            let this = unsafe { &mut *(self as *mut _ as *mut RequestBuf<Option<&'out T>>) };
```

do you wanna try applying these changes in https://github.com/yaahc/nostd-error-poc/ first and open a PR there, once it's merged there I'll copy the changes into the RFC.

    }

    /// Get the `TypeId` of the requested type.

>     fn type_id(&self) -> TypeId {

Reviewer 4: Might be worth removing the method completely and accessing `self.type_id` at any call-sites.
